helium-lib/src/token.rs

```diff
@@ -113,27 +121,79 @@ pub async fn balance_for_addresses(
         .await
 }
 
-pub async fn pyth_price(settings: &Settings, token: Token) -> Result<pyth_sdk_solana::Price> {
-    let price_key = token
-        .price_key()
-        .ok_or_else(|| DecodeError::other(format!("No pyth price key for {token}")))?;
-    let client = settings.mk_solana_client()?;
-    let mut price_account = client.get_account(price_key).await?;
-    let price_feed =
-        pyth_sdk_solana::state::SolanaPriceAccount::account_to_feed(price_key, &mut price_account)?;
-
-    use std::time::{SystemTime, UNIX_EPOCH};
-    let current_time = SystemTime::now().duration_since(UNIX_EPOCH)?;
-    price_feed
-        .get_ema_price_no_older_than(
-            current_time
-                .as_secs()
-                .try_into()
-                .map_err(DecodeError::from)?,
-            10 * 60,
-        )
-        .ok_or_else(|| DecodeError::other("No token price found"))
-        .map_err(Error::from)
+pub mod price {
+    use super::*;
+    use pyth_solana_receiver_sdk::price_update::{self, PriceUpdateV2};
+    use rust_decimal::prelude::*;
+
+    pub use pyth_solana_receiver_sdk::price_update::FeedId;
+    pub const DC_PER_USD: i64 = 100_000;
+
+    #[derive(Debug, thiserror::Error)]
+    pub enum PriceError {
+        #[error("invalid or unsupported token: {0}")]
+        InvalidToken(super::Token),
+        #[error("invalid price feed")]
+        InvalidFeed,
+        #[error("price too old")]
+        TooOld,
+        #[error("price below 0")]
+        Negative,
+        #[error("invalid price timestamp: {0}")]
+        InvalidTimestamp(i64),
+        #[error("unsupported positive price exponent")]
+        PositiveExponent,
+    }
+
+    #[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+    pub struct Price {
+        pub timestamp: DateTime<Utc>,
+        pub price: Decimal,
+        pub token: super::Token,
+    }
+
+    pub fn feed_from_hex(str: &str) -> Result<FeedId> {
+        let feed_id =
+            price_update::get_feed_id_from_hex(str).map_err(|_| PriceError::InvalidFeed)?;
+        Ok(feed_id)
+    }
+
+    pub async fn get(settings: &Settings, token: Token) -> Result<Price> {
```

Comment:
Given this is a library that we intend to use other places, shouldn't this function take an already built RpcClient that way it is on the caller how its want's to manage the connection?

Reply:
settings is basically just a url, but yes, I'll change this one to take a configured client.

Ideally, there'd be a `TryInto<SolanaRpcClient` as the first param but I can't make that work given how clients are not Send/Clone and one is passed by reference vs by value